CFE-951: Dealing with permissions that used to be handled by rxdirs default of true

nickanderson opened the pull request.

vpodzime commented:
jenkins: [![Build Status](https://ci.cfengine.com/buildStatus/icon?job=pr-pipeline&build=7686)](https://ci.cfengine.com/job/pr-pipeline/7686/)

vpodzime commented:
This supersedes #2264

nickanderson commented:
@cf-bottom let's see what Jenkins has to say about this now.

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=7691)](https://ci.cfengine.com//job/pr-pipeline/7691/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7691/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7691/

craigcomstock commented on cfe_internal/enterprise/CFE_hub_specific.cf:
```diff
@@ -765,27 +765,47 @@ bundle agent cfe_internal_enterprise_policy_analyzer
+    enterprise_edition.policy_server::
 
-      "$(analyzer_dir)/."
+      "$(cfe_internal_hub_vars.docroot)/analyzer/."
```
Why the change here and not on line 785 below?

nickanderson replied:
Yeah, it got me for a few minutes as well.

Because analyzer_dir != docroot/analyzer.  Analyzer_dir is in opt, this docroot analyzer path is a directory where the web server writes the enablement flag file if someone clicks that button in MP. Then analyzer_dir is where the MPF is copied and where the webserver reads the policy from for rendering in MP.

craigcomstock replied:
So this is really a bug-fix of sorts. Ack.

nickanderson replied:
MMMmm, I don't think so, I think that previously the web server had the ability to create the directory in the docroot if it didn't exist.

craigcomstock commented on cfe_internal/enterprise/CFE_knowledge.cf:
```diff
@@ -51,17 +51,50 @@ bundle agent cfe_internal_setup_knowledge
-      file_select => cfe_internal_docroot_perms,
-      perms => mog("0460", "root", $(def.cf_apache_group) ),
-      depth_search => recurse_exclude("inf");   # see exclude dirs in recurse_exclude() body
+      # "$(cfe_internal_hub_vars.docroot)"
```
delete this section instead of commenting out.

craigcomstock commented on cfe_internal/enterprise/CFE_knowledge.cf:
```diff
@@ -177,11 +235,14 @@ bundle agent cfe_internal_setup_knowledge
         depth_search => recurse( "inf" ),
         perms => mog("0600", $(def.cf_apache_user), $(def.cf_apache_group));
 
+      # TODO FIXUP peerms in vendor/codeigniter/framework/system/core/Log.php
```
handle TODO and link the PR to this one yeah?

craigcomstock commented on lib/files.cf:
```diff
@@ -1937,6 +1937,25 @@ body file_select exclude(name)
+body file_select not_dir
+# @brief Select all files that are not directories
+{
+      # Select all kinds of files except for directories
```
delete commented out bits.

nickanderson replied:
yeaap

nickanderson commented:
> seems like this should be linked to mission-portal and buildscripts PRs.
> 
> mission-portal for one TODO buildscripts to enforce/apply these permission changes during packaging?

I think that buildscripts PRs would be separate. Here I am just aiming for compatibility with how rxdirs behaves before and after the change in default.

So, if any of these perms are not correct in buildscripts today, it's something that is already fixed during the first policy run

nickanderson commented:
@cf-bottom let's see what Jenkins has to say about this now.

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=7701)](https://ci.cfengine.com//job/pr-pipeline/7701/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7701/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7701/

nickanderson commented:
@cf-bottom can you please have jenkins test this?

nickanderson commented:
@cf-bottom can you please have jenkins test this?

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=7709)](https://ci.cfengine.com//job/pr-pipeline/7709/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7709/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7709/

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=7710)](https://ci.cfengine.com//job/pr-pipeline/7710/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7710/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7710/

nickanderson commented:
@cf-bottom I killed my previous builds, can you run this ONE through jenkins please?

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=7711)](https://ci.cfengine.com//job/pr-pipeline/7711/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7711/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7711/

nickanderson commented:
@cf-bottom, again I killed my builds and pushed a new commit. please ask jenkins to test this for me

nickanderson commented:
@cf-bottom, jenkins please

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=7715)](https://ci.cfengine.com//job/pr-pipeline/7715/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7715/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7715/

nickanderson commented:
@cf-bottom, jenkins please

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=7716)](https://ci.cfengine.com//job/pr-pipeline/7716/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7716/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7716/

vpodzime commented:
sequential-tests retry: [![Build Status](https://ci.cfengine.com/buildStatus/icon?job=sequential-tests&build=3555)](https://ci.cfengine.com/job/sequential-tests/3555/)

nickanderson commented on cfe_internal/enterprise/CFE_hub_specific.cf:
```diff
@@ -462,6 +462,17 @@ body classes ENT_5279
+        policy => "present",
+        description => "CFEngine PostgreSQL",
+        group_primary => "cfpostgres",
+        home_dir => "/home/cfpostgres",
```
Where should we place this? `/var/cfengine/pg/home`, `/opt/cfengine/home/cfpostgres` ?

vpodzime replied:
Nowhere, I don't think the *cfpostgres* user needs a home directory. The passing *FR-tests* in https://ci.cfengine.com/job/pr-pipeline/7716/ contain things like:
```
00:04:59 + ssh -o LogLevel=ERROR -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o ConnectTimeout=60 root@54.229.114.26 'su - cfpostgres --command "PGDATA=/var/cfengine/state/pg/data /var/cfengine/bin/pg_ctl status"'
00:05:00 su: warning: cannot change directory to /home/cfpostgres: No such file or directory
00:05:00 could not change directory to "/root": Permission denied
```
So it's noise, but harmless (in our cases). The proper thing to do is to `cd /tmp` first in all those commands which removes the second warning, the first one will go away when there's no home directory, AFAICT.

nickanderson replied:
NOTE TO SELF: This means, for all commands promises that use `psql` we need to add/extend a `contain` body that uses `chdir => "/tmp".

vpodzime commented on cfe_internal/enterprise/CFE_hub_specific.cf:
```diff
@@ -765,27 +776,47 @@ bundle agent cfe_internal_enterprise_policy_analyzer
+                           "that is written by Mission Portal when someone",
+                           "clicks to enable the feature. Thus, it needs to be",
+                           "readable, writeable and executable web",
+                           "server.");
```
Missing "***for the** web server*"? Also, missing spaces at ends or beginnings of lines.

nickanderson replied:
ACK

vpodzime commented on cfe_internal/enterprise/CFE_knowledge.cf:
```diff
@@ -51,17 +51,43 @@ bundle agent cfe_internal_setup_knowledge
+      comment => "The top level docroot needs to be readable and executable by the web server.",
       handle => "cfe_internal_setup_knowledge_dir_doc_root",
-      perms => mog("0460", "root", $(def.cf_apache_group) );
+      perms => mog("0450", "root", $(def.cf_apache_group) );
```
It doesn't make much sense to give the *root* user read access but not execute. Strictly speaking this would mean even the agent process could not access and modify files in these directories.

nickanderson replied:
ACK, mog( 0550 )

vpodzime commented on cfe_internal/enterprise/CFE_knowledge.cf:
```diff
@@ -466,13 +538,15 @@ body depth_search recurse_basedir_exclude(d)
 
 body file_select cfe_internal_docroot_perms
-# @brief Select files not named `.htacces` or `settings.ldap.php`
+# @brief Select files (not dirs) not named `.htacces` or `settings.ldap.php`
```
Might as well fix the *.htacces**s*** name.

nickanderson replied:
ACK.

vpodzime commented on cfe_internal/enterprise/CFE_knowledge.cf:
```diff
@@ -466,13 +538,15 @@ body depth_search recurse_basedir_exclude(d)
       # settings.ldap.php permissions are handled explicitly in it's own bundle
       # ha_enabled permissions is handled explicitly by cfengine_enterprise_ha_enabled_semaphore_present
-  file_result => "!leaf_name";
+  file_result => "!leaf_name.!file_types.!path_name";
```
Wow, so much negativity!

vpodzime reviewed (commented):
I know this is a lot of painful work, but I must say I really like the result(s)! THANKS! :clap:

nickanderson commented:
@cf-bottom jenkins, please

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=7721)](https://ci.cfengine.com//job/pr-pipeline/7721/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7721/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7721/

craigcomstock commented on cfe_internal/enterprise/CFE_knowledge.cf:
```diff
@@ -51,17 +51,43 @@ bundle agent cfe_internal_setup_knowledge
+        handle => "cfe_internal_setup_knowledge_dir_doc_root_vendor_dirs",
+        depth_search => recurse_with_base("inf"),
+        file_select => dirs,
+        perms => mog("0450", "root", $(def.cf_apache_group) );
```
oh, right, exec aka 'x' for dirs :+1:

craigcomstock replied:
maybe mention that in the comment, that x dirs is needed and separate from perms for files which comes next.

nickanderson replied:
ACK, Improved comments.

craigcomstock commented on cfe_internal/enterprise/CFE_knowledge.cf:
```diff
@@ -51,17 +51,43 @@ bundle agent cfe_internal_setup_knowledge
+        perms => mog("0450", "root", $(def.cf_apache_group) );
+
+      "$(cfe_internal_hub_vars.public_docroot)/images/." -> { "CFE-951" }
+        comment => "The vendor directory contains dependencies from the code ignitor framework",
```
wrong comment? this is images not vendor dir right?

nickanderson replied:
ACK

craigcomstock commented on cfe_internal/enterprise/CFE_knowledge.cf:
```diff
@@ -51,17 +51,43 @@ bundle agent cfe_internal_setup_knowledge
+        perms => mog("0440", "root", $(def.cf_apache_group) );
+
+      "$(cfe_internal_hub_vars.public_docroot)/themes/." -> { "CFE-951" }
+        comment => "The vendor directory contains dependencies from the code ignitor framework",
```
wrong comment, themes, not vendor/code-ignitor?

nickanderson replied:
ACK

nickanderson commented on cfe_internal/enterprise/federation/federation.cf:
```diff
@@ -616,9 +642,9 @@ bundle agent imported_data
+      "/bin/bash" -> { "CFE-951" }
         arglist => {"$(cfengine_enterprise_federation:config.bin_dir)/import.sh"},
-        contain => default:in_shell;
+        contain => default:in_shell_dir( "/tmp" );
```
@craigcomstock in what directory is import.sh typically run from?

In the script I see that it's outputting information to things like schema_setup.log but I don't see any changes to the working directory, maybe I am just missing them.

craigcomstock replied:
Right, this is something I was fixing yesterday in fact. It is just written to cwd which I can confirm on a superhub I have handy. But the scripts really should write to a specific location, which is what I was fixing yesterday. Changing to `/tmp` is fine.

nickanderson replied:
Yeah, i was actually asking about this in relation to something completely different (not this PR).

craigcomstock replied:
I will likely submit a PR "fixing" this "unexplicit" location of these log files.

craigcomstock commented on cfe_internal/enterprise/CFE_knowledge.cf:
```diff
@@ -553,9 +630,15 @@ body depth_search cfe_internal_docroot_application_perms
+body perms state_dir_system_owned_files
 {
-      rxdirs => "true";
+#+begin_ENT-951
```
remove org mode stuff here?

nickanderson replied:
nay.

craigcomstock commented on cfe_internal/enterprise/federation/federation.cf:
```diff
@@ -317,8 +326,10 @@ bundle agent clean_when_off
-    "remote_hubs_table_row_count" string => execresult(`$(sys.bindir)/psql cfsettings --quiet --tuples-only --command "SELECT COUNT(*) FROM remote_hubs" 2>/dev/null`, useshell);
-    "federated_reporting_settings_table_row_count" string => execresult(`$(sys.bindir)/psql cfsettings --quiet --tuples-only --command "SELECT COUNT(*) FROM federated_reporting_settings" 2>/dev/null`, useshell);
+      "remote_hubs_table_row_count" -> { "CFE-951" }
+        string => execresult(`cd /tmp; $(sys.bindir)/psql cfsettings --quiet --tuples-only --command "SELECT COUNT(*) FROM remote_hubs" 2>/dev/null`, useshell);
```
why `cd tmp` here since you are not containing it with `cfpostgres` user right? It "just works" as root I think because we have a root role.

https://github.com/cfengine/buildscripts/blob/master/packaging/common/cfengine-hub/postinstall.sh#L775-L777

```
    if ! echo "$db_user_list" | grep root >/dev/null; then
      su cfpostgres -c "$PREFIX/bin/createuser -d -s -w root"
    fi
```

nickanderson replied:
ack

craigcomstock commented on cfe_internal/enterprise/federation/federation.cf:
```diff
@@ -345,10 +356,13 @@ bundle agent clean_when_off
-        if => isgreaterthan( "$(remote_hubs_table_row_count)", "0" );
-      `$(sys.bindir)/psql cfsettings --quiet --command "TRUNCATE TABLE federated_reporting_settings"` -> { "ENT-7233" }
-        if => isgreaterthan( "$(federated_reporting_settings_table_row_count)", "0" );
+      `$(sys.bindir)/psql cfsettings --quiet --command "TRUNCATE TABLE remote_hubs"` -> { "ENT-7233", "CFE-951" }
```
Again, you aren't containing as `cfpostgres` user and I don't think you need to as the root postgres role/user can do these operations because they are given "superuser" privileges as mentioned in the above buildscripts post install script content.

craigcomstock commented on cfe_internal/enterprise/federation/federation.cf:
```diff
@@ -565,9 +590,9 @@ bundle agent exported_data
+      "/bin/bash" -> { "CFE-951" }
         arglist => {"$(cfengine_enterprise_federation:config.bin_dir)/dump.sh"},
-        contain => default:in_shell,
+        contain => default:in_shell_dir("/tmp"),
```
why in `/tmp` ? isn't this run as root here?

nickanderson replied:
WHY? Because biggest hammer.
It is run as root.
ACK,

craigcomstock commented on cfe_internal/enterprise/federation/federation.cf:
```diff
@@ -616,9 +642,9 @@ bundle agent imported_data
+      "/bin/bash" -> { "CFE-951" }
         arglist => {"$(cfengine_enterprise_federation:config.bin_dir)/import.sh"},
-        contain => default:in_shell;
+        contain => default:in_shell_dir( "/tmp" );
```
why in `/tmp`? ;)

craigcomstock commented on cfe_internal/enterprise/federation/federation.cf:
```diff
@@ -627,12 +653,13 @@ bundle agent superhub_schema
                    },
-        classes => psql_wrapper_exit_codes;
+        classes => psql_wrapper_exit_codes,
+        contain => default:in_dir( "/tmp" );
```
no need for `/tmp`?

nickanderson replied:
Well, inside psql_wrapper.sh.mutache I see `su - cfpostgres`.

```
OUT=$(su - cfpostgres --command "{{{vars.sys.bindir}}}/psql --quiet --tuples-only --no-align --no-psqlrc \"$1\" --command=\"$2\"" 2> $TMP)
```

craigcomstock replied:
I would say DRY tells us we should `cd /tmp` inside the psql_wrapper.sh.mustache file and leave it out of policy. But it's only two calls from policy anyway so doesn't get us that much. :)

nickanderson replied:
![image](https://user-images.githubusercontent.com/202896/158444985-4c0fffb7-8459-46dd-bc73-43a012e797b8.png)

nickanderson replied:
ACK.

craigcomstock commented on cfe_internal/enterprise/federation/federation.cf:
```diff
@@ -645,13 +672,14 @@ bundle agent ensure_feeders
         classes => psql_wrapper_exit_codes,
-        if => isgreaterthan(length(feeders), 0);
+        if => isgreaterthan(length(feeders), 0),
+        contain => default:in_dir( "/tmp" );
```
no need for `/tmp`?

nickanderson replied:
Similar to https://github.com/cfengine/masterfiles/pull/2265#discussion_r827273694

inside psql_wrapper.sh.mutache I see `su - cfpostgres`.

```
OUT=$(su - cfpostgres --command "{{{vars.sys.bindir}}}/psql --quiet --tuples-only --no-align --no-psqlrc \"$1\" --command=\"$2\"" 2> $TMP)
```

craigcomstock replied:
ditto, makes sense to me to put the `cd /tmp` in the psql_wrapper script instead of in policy.

craigcomstock commented on lib/cfe_internal_hub.cf:
```diff
@@ -133,11 +133,11 @@ WHERE  z.hostkey = t.hostkey
         inform => "false",
 @endif
-        contain => silent,
+        contain => default:silent_in_dir( "/tmp" ),
```
no need, as this is run as root.

nickanderson replied:
ack

craigcomstock commented on lib/cfe_internal_hub.cf:
```diff
@@ -149,12 +149,12 @@ bundle agent cfe_internal_database_cleanup_cf_hub_status (row_count)
-      "has_sql_function_cleanup_historical_data"
-        expression => returnszero( "$(sys.bindir)/psql cfdb -c \"SELECT 'cleanup_historical_data'::regproc;\" > /dev/null 2>&1", useshell ),
+      "has_sql_function_cleanup_historical_data" -> { "CFE-951" }
+        expression => returnszero( "cd /tmp; $(sys.bindir)/psql cfdb -c \"SELECT 'cleanup_historical_data'::regproc;\" > /dev/null 2>&1", useshell ),
```
no need to cd tmp, running as root should "just work"

craigcomstock commented on lib/cfe_internal_hub.cf:
```diff
@@ -187,18 +187,18 @@ bundle agent cfe_internal_database_cleanup_cf_hub_status (row_count)
         inform => "false",
 @endif
-        contain => silent,
+        contain => default:silent_in_dir( "/tmp"),
```
rm

craigcomstock commented on lib/cfe_internal_hub.cf:
```diff
@@ -212,11 +212,11 @@ bundle agent cfe_internal_database_cleanup_diagnostics (settings)
         inform => "false",
 @endif
-        contain => silent,
+        contain => default:silent_in_dir( "/tmp"),
```
rm

nickanderson replied:
ACK

craigcomstock reviewed (changes requested):
Looks good. A few comment fixes and generally I don't think we need the `/tmp` business when running `psql` as root as we are in all these cases. Root has superuser privs in postgres so should "just work" as-is.

nickanderson commented:
> Looks good. A few comment fixes and generally I don't think we need the `/tmp` business when running `psql` as root as we are in all these cases. Root has superuser privs in postgres so should "just work" as-is.

I addressed your comments specifically, but I didn't go and revert the /tmp change, instead addressed each you commented on. I saw a couple you didn't comment on that I took back, but there are possibly more, i havnet looked  yet.

craigcomstock reviewed (commented):
all looks good, squash, test, green => merge I'd say 👍 🏆

craigcomstock commented on lib/cfe_internal_hub.cf:
```diff
@@ -232,18 +232,18 @@ bundle agent cfe_internal_database_cleanup_promise_log (history_length_days)
         inform => "false",
 @endif
-        contain => default:silent_in_dir( "/tmp"),
+        contain => default:silent,
```
👍

craigcomstock commented on lib/cfe_internal_hub.cf:
```diff
@@ -232,18 +232,18 @@ bundle agent cfe_internal_database_cleanup_promise_log (history_length_days)
         inform => "false",
 @endif
-        contain => default:silent_in_dir( "/tmp"),
+        contain => default:silent,
```
👍

craigcomstock commented on lib/cfe_internal_hub.cf:
```diff
@@ -258,11 +258,11 @@ bundle agent cfe_internal_database_partitioning()
         inform => "false",
 @endif
-        contain => default:silent_in_dir( "/tmp"),
+        contain => default:silent,
```
👍

nickanderson commented:
> all looks good, squash, test, green => merge I'd say +1 trophy

![image](https://user-images.githubusercontent.com/202896/158449127-08f0a48e-5876-44f0-8dc9-86a0cce32b60.png)

nickanderson commented:
@cf-bottom jenkins, please

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=7726)](https://ci.cfengine.com//job/pr-pipeline/7726/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7726/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7726/

vpodzime commented:
This seems to be the last remaining bit:
```
+ curl --fail --insecure --include --silent -X POST -u admin:testingFR https://54.217.64.24/index.php/hub_management/api/addFeeder -H "content-type: multipart/form-data;" -F api_url=https://3.250.95.216/ -F userName=admin -F password=testingFR -F enabled=true -F target_state=on -F skipSslVerification=true -F "transport_method[value]=pull_over_rsync"
HTTP/1.1 100 Continue
++ failure
```

vpodzime commented:
Info from @aleksandrychev:
```
file_put_contents(/opt/cfengine/federation/cfapache/setup-status.json): failed to open stream: Permission denied
drw-rw---x 3 root     root     4096 Mar 16 11:51 federation
```

nickanderson commented:
> Info from @aleksandrychev:
> 
> ```
> file_put_contents(/opt/cfengine/federation/cfapache/setup-status.json): failed to open stream: Permission denied
> drw-rw---x 3 root     root     4096 Mar 16 11:51 federation
> ```

![hunting-for-wabbits](https://user-images.githubusercontent.com/202896/158594912-bdad51cf-3470-45d4-92e3-7607c34bbbde.gif)

vpodzime commented:
```
 ls -l /opt/cfengine/
total 16
drwxr-xr-x 2 cfapache cfapache 4096 Mar 16 11:50 dc-scripts
drw-rw---x 6 root     root     4096 Mar 16 12:02 federation
```

nickanderson commented:
@cf-bottom jenkins, please

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=7729)](https://ci.cfengine.com//job/pr-pipeline/7729/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7729/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7729/

nickanderson commented:
@cf-bottom jenkins, again please

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=7731)](https://ci.cfengine.com//job/pr-pipeline/7731/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7731/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7731/

vpodzime commented:
Close!
```
    info: Executing "no timeout,uid=1001,gid=1001" ... "/bin/bash /opt/cfengine/federation/bin/transport.sh pull 54.170.207.210 18.202.80.9 176.34.204.221 3.250.113.76 3.249.30.211 52.50.29.14 3.248.184.6"
   error: Finished command related to promiser "/bin/bash" -- an error occurred, returned 126
  notice: Q: ".../bin/bash /opt/": /bin/bash: /opt/cfengine/federation/bin/transport.sh: Permission denied
    info: Last 1 quoted lines were generated by promiser "/bin/bash /opt/cfengine/federation/bin/transport.sh pull 54.170.207.210 18.202.80.9 176.34.204.221 3.250.113.76 3.249.30.211 52.50.29.14 3.248.184.6"
    info: Completed execution of "/bin/bash /opt/cfengine/federation/bin/transport.sh pull 54.170.207.210 18.202.80.9 176.34.204.221 3.250.113.76 3.249.30.211 52.50.29.14 3.248.184.6"
   error: Method "data_transport" failed in some repairs
```

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=7733)](https://ci.cfengine.com//job/pr-pipeline/7733/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7733/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7733/

nickanderson commented:
@cf-bottom jenkins, again please

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=7736)](https://ci.cfengine.com//job/pr-pipeline/7736/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7736/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7736/

nickanderson commented:
dont cross all your xes and dot your ts

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=7737)](https://ci.cfengine.com//job/pr-pipeline/7737/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7737/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7737/

nickanderson commented:
Pretty close. The failure is in triple upgrade test, which perhaps is an un-releated failure.

nickanderson commented:
Ok, tidied up some promisees and squahsed.

@cf-bottom jenkins, one more time, we're gonna celebrate

https://www.youtube.com/watch?v=FGBhQbmPwH8

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=7738)](https://ci.cfengine.com//job/pr-pipeline/7738/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/7738/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-7738/

vpodzime reviewed (approved):
Looks good to me. And it's too good to be left unmerged.

vpodzime commented:
@craigcomstock feel free to re-review this and we can address issues in a follow-up. I'm merging this now to get greener builds.
